File: back/src/main/java/com/back/domain/book/service/BookRankService.java
package com.back.domain.book.service;

import com.back.domain.book.dto.BookDetailDto;
import com.back.domain.book.dto.BookDto;
import com.back.domain.book.entity.Book;
import com.back.domain.book.repository.BookRepository;
import com.back.domain.member.entity.Member;
import com.back.domain.review.entity.Review;
import com.back.domain.review.repository.ReviewRepository;
import com.back.domain.wish.repository.WishRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.NoSuchElementException;

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class BookRankService {
    private final BookRepository bookRepository;
    private final ReviewRepository reviewRepository;

    public List<BookDto> getBooksByRating(int page, int size) {
        Pageable pageable = PageRequest.of(page, size);
        return bookRepository.findAllByOrderByAverageRatingDesc(pageable)
                .stream()
                .map(BookDto::new)
                .toList();
    }

    public List<BookDto> getBooksByReviewCnt(int page, int size) {
        Pageable pageable = PageRequest.of(page, size);

        return bookRepository.findAllByOrderByReviewCountDesc(pageable)
                .stream()
                .map(BookDto::new)
                .toList();
    }

}

File: back/src/main/java/com/back/domain/book/service/BookRankService.kt
package com.back.domain.book.service

import com.back.domain.book.entity.Book
import com.back.domain.book.repository.BookRepository
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest

@org.springframework.stereotype.Service
@org.springframework.transaction.annotation.Transactional(readOnly = true)
class BookRankService(
    private val bookRepository: BookRepository
) {

    fun getBooksOrderByRating(page: Int, size: Int): Page<Book> {
        val pageable = PageRequest.of(page, size)
        return bookRepository.findAllByOrderByAverageRatingDesc(pageable)
    }

    fun getBooksOrderByReviewCnt(page: Int, size: Int): Page<Book> {

        val pageable = PageRequest.of(page, size)

        return bookRepository.findAllByOrderByReviewCountDesc(pageable)
    }
}
